modules/core/src/main/java/org/apache/ignite/internal/thread/PublicApiThreading.java

```diff
@@ -48,12 +51,16 @@ public static void endInternalCall() {
      * @return Call result.
      */
     public static <T> T doInternalCall(Supplier<T> call) {
+        boolean wasInInternalCallBefore = inInternalCall();
+
         startInternalCall();
```

Comment:
Do we need to check the `wasInInternalCallBefore` flag? I understand that this method is idempotent, but it will be more consistent. Maybe we should even avoid the try-finally schenanigans if we are already in an internal call.